### PR TITLE
bakery: add Oven and Bakery types

### PR DESCRIPTION
**Pull request by rogpeppe**
The Oven is the type responsible for creating macaroons and verifying
their signatures. Responsibility for checking caveats and operations
is left to the Checker. The Oven should be compatible with old macaroon
ids too - if an old id is found, it is assumed to be associated with the
LoginOp operation, as prior to this all macaroons have been essentially
treated as login macaroons.

We use protocol buffers to encode the new macaroon identifiers,
which gives us a nicely compact format that we don't have to invent
ourselves - note that macaroon ids are considered opaque
so only the service generating them will need to be able to read them,
so we don't need to worry about python and javascript support for
protocol buffers.

Although technically both Oven and Checker are independently useful,
many applications will want to combine them, so we make a new transparent
Bakery type that puts them together for convenience. Some might argue
that Bakery could embed both types rather than just as fields - I don't
mind either way.

I've marked the examples to be ignored for now so that they don't clutter
this already huge PR.  To be good examples, they will be better off
rewritten to take advantage of the new API rather than just patched up.

We also rename the word "storage" to "store" throughout as there
is no good reason for the distinction - it's only confusing.

### CODE REVIEW

File: bakery/bakery.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+	// RootKeyStore holds the root key store to use. If you need to
+	// use a different root key store for different operations,
+	// you'll need to pass a RootKeyStoreForOps value to NewOven
+	// directly.
+	//
+	// If this is nil, New will use NewMemRootKeyStore().
+	RootKeyStore RootKeyStore
+
+	// Locator is used to find third party public keys when adding
```
Comment (mhilton) [inline bakery/bakery.go]: This is not all that it's used for, it also gets the caveat ID encoding version.

File: bakery/bakery.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	// third party caveats.
+	//
+	// If this is nil, the Oven will not able to add any third party
+	// caveats.
+	Locator ThirdPartyLocator
+
+	// Key holds the private key of the  If this is nil,
```
Comment (mhilton) [inline bakery/bakery.go]: of the what?

File: bakery/example/targetservice.go [hunk 3]
```diff
@@ -61,7 +63,7 @@ func (srv *targetServiceHandler) serveGold(w http.ResponseWriter, req *http.Requ
 }
 
 func (srv *targetServiceHandler) serveSilver(w http.ResponseWriter, req *http.Request) {
-	ctxt := checkers.ContextWithOperations(context.TODO(), "silver")
+	ctxt := checkers.ContextWithOperations(context.TODO(), "gold")
```
Comment (mhilton) [inline bakery/example/targetservice.go]: Why have you changed this? it seems wrong now.

File: bakery/oven.go [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+	}
+	return nil
+}
+
+// Key returns the oven's private/public key par.
+func (o *Oven) Key() *KeyPair {
+	return o.p.Key
+}
+
+func (o *Oven) Locator() ThirdPartyLocator {
```
Comment (mhilton) [inline bakery/oven.go:264]: doc comment :)

File: bakery/oven.go [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+		Nonce:     nonce,
+		StorageId: storageId,
+		Ops: []*macaroonpb.Op{{
+			Entity:  entity,
+			Actions: []string{"*"},
+		}},
+	}, nil
+}
+
+// newMultiEntity returns a new multi-op entity name that represents
```
Comment (mhilton) [inline bakery/oven.go]: newOpsEntity

File: httpbakery/checkers.go [hunk 6]
```diff
@@ -12,9 +12,6 @@ import (
 
 type httpRequestKey struct{}
 
-// ContextWithRequest returns the context with information from the
```
Comment (mhilton) [inline httpbakery/checkers.go:15]: Why've you removed this? it still seems true

File: httpbakery/checkers.go [hunk 7]
```diff
@@ -25,16 +22,10 @@ func requestFromContext(ctxt context.Context) *http.Request {
 }
 
 const (
-	// CondClientIPAddr holds the first party caveat condition
```
Comment (mhilton) [inline httpbakery/checkers.go:28]: And here

File: httpbakery/checkers_test.go [hunk 8]
```diff
@@ -149,7 +149,7 @@ var checkerTests = []struct {
 	},
 	checks: []checkTest{{
 		caveat:      httpbakery.ClientOriginCaveat(""),
-		expectError: `caveat "http:origin" not satisfied: request has invalid Origin header; got "somewhere"`,
+		expectError: `caveat "http:origin " not satisfied: request has invalid Origin header; got "somewhere"`,
```
Comment (mhilton) [inline httpbakery/checkers_test.go]: This just seems wrong, why is there a trailing space?

File: httpbakery/checkers.go [hunk 9]
```diff
@@ -43,18 +34,6 @@ var allCheckers = map[string]checkers.Func{
 }
 
 // RegisterCheckers registers all the HTTP checkers with the given checker.
-// Current checkers include:
```
Comment (mhilton) [inline httpbakery/checkers.go:46]: I'd leave this in, it looks useful.

**Review by mhilton — APPROVED** [review]
Great stuff, some small nits

File: bakery/bakery.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	// Checker holds the checker used to check first party caveats.
+	// If this is nil, New will use checkers.New(nil).
+	Checker FirstPartyCaveatChecker
+
+	// RootKeyStore holds the root key store to use. If you need to
+	// use a different root key store for different operations,
+	// you'll need to pass a RootKeyStoreForOps value to NewOven
+	// directly.
+	//
+	// If this is nil, New will use NewMemRootKeyStore().
```
Comment (alesstimec) [inline bakery/bakery.go:32]: a statement saying that for production, this should never be nil.. or something to that effect..

File: bakery/bakery.go (hunk 2 above)
Comment (alesstimec) [reply bakery/bakery.go]: of the the..  ;)

File: bakery/mgorootkeystore/rootkey.go [hunk 11]
```diff
@@ -1,4 +1,4 @@
-package mgostorage
+package mgorootkeystore
```
Comment (alesstimec) [inline bakery/mgorootkeystore/rootkey.go:3]: no package godoc?

File: bakery/oven.go [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+		// It's an old-style id, probably with a hyphenated UUID.
+		// so trim that off.
+		if i := bytes.LastIndexByte(id, '-'); i >= 0 {
+			storageId = id[0:i]
+		}
+	}
+	return storageId, []Op{LoginOp}, nil
+}
+
+// c. takes a macaroon with the given version from the oven, associates it with the given operations
```
Comment (alesstimec) [inline bakery/oven.go]: c. ?

File: bakery/service.go [hunk 13]
```diff
@@ -608,29 +260,29 @@ type ThirdPartyCaveatInfo struct {
...
 //
 // If the caveat kind was not recognised, the checker should return an
 // error with a ErrCaveatNotRecognized cause.
-type ThirdPartyChecker interface {
+type ThirdPartyCaveatChecker interface {
 	CheckThirdPartyCaveat(ctxt context.Context, info *ThirdPartyCaveatInfo) ([]checkers.Caveat, error)
 }
 
-type ThirdPartyCheckerFunc func(context.Context, *ThirdPartyCaveatInfo) ([]checkers.Caveat, error)
+type ThirdPartyCaveatCheckerFunc func(context.Context, *ThirdPartyCaveatInfo) ([]checkers.Caveat, error)
```
Comment (alesstimec) [inline bakery/service.go:276]: godoc

File: bakery/service.go [hunk 14]
```diff
@@ -608,29 +260,29 @@ type ThirdPartyCaveatInfo struct {
...
 
-func (c ThirdPartyCheckerFunc) CheckThirdPartyCaveat(ctxt context.Context, info *ThirdPartyCaveatInfo) ([]checkers.Caveat, error) {
+func (c ThirdPartyCaveatCheckerFunc) CheckThirdPartyCaveat(ctxt context.Context, info *ThirdPartyCaveatInfo) ([]checkers.Caveat, error) {
```
Comment (alesstimec) [inline bakery/service.go:280]: godoc

File: bakery/service.go [hunk 15]
```diff
@@ -640,3 +292,13 @@ type FirstPartyCaveatChecker interface {
 	// caveat checker.
 	Namespace() *checkers.Namespace
 }
+
+type FirstPartyCaveatCheckerFunc func(ctxt context.Context, caveat string) error
```
Comment (alesstimec) [inline bakery/service.go]: godoc here and the two methods below

File: httpbakery/keyring.go [hunk 16]
```diff
@@ -41,8 +41,6 @@ type ThirdPartyLocator struct {
 	cache         *bakery.ThirdPartyStore
 }
 
-var _ bakery.ThirdPartyLocator = (*ThirdPartyLocator)(nil)
```
Comment (alesstimec) [inline httpbakery/keyring.go:44]: why?

**Review by alesstimec — CHANGES_REQUESTED** [review]
a few comments and questions.. otherwise LGTM

File: bakery/bakery.go (hunk 1 above)
Comment (rogpeppe) [reply bakery/bakery.go]: fixed

File: bakery/bakery.go (hunk 2 above)
Comment (rogpeppe) [reply bakery/bakery.go]: done

File: bakery/example/targetservice.go (hunk 3 above)
Comment (rogpeppe) [reply bakery/example/targetservice.go]: done

File: bakery/oven.go (hunk 4 above)
Comment (rogpeppe) [reply bakery/oven.go:264]: done

File: bakery/oven.go (hunk 5 above)
Comment (rogpeppe) [reply bakery/oven.go]: done

File: httpbakery/checkers.go (hunk 6 above)
Comment (rogpeppe) [reply httpbakery/checkers.go:15]: reverted all

File: httpbakery/checkers_test.go (hunk 8 above)
Comment (rogpeppe) [reply httpbakery/checkers_test.go]: reverted (and it caused a test failure too)

File: bakery/bakery.go (hunk 10 above)
Comment (rogpeppe) [reply bakery/bakery.go:32]: done

File: bakery/mgorootkeystore/rootkey.go (hunk 11 above)
Comment (rogpeppe) [reply bakery/mgorootkeystore/rootkey.go:3]: done

File: bakery/oven.go (hunk 12 above)
Comment (rogpeppe) [reply bakery/oven.go]: fixed

File: bakery/service.go (hunk 13 above)
Comment (rogpeppe) [reply bakery/service.go:276]: done

File: bakery/service.go (hunk 14 above)
Comment (rogpeppe) [reply bakery/service.go:280]: done

File: bakery/service.go (hunk 15 above)
Comment (rogpeppe) [reply bakery/service.go]: removed the type.

File: httpbakery/keyring.go (hunk 16 above)
Comment (rogpeppe) [reply httpbakery/keyring.go:44]: because I got it wrong. :)
reinstated.

**Pull request merged**